include/extractor/segment_data_container.hpp

```diff
@@ -0,0 +1,157 @@
+    }
+    auto& ForwardWeight(const DirectionalGeometryID id, const SegmentOffset offset)
+    {
+        return fwd_weights[index[id] + 1 + offset];
+    }
+    auto& ReverseWeight(const DirectionalGeometryID id, const SegmentOffset offset)
+    {
+        return rev_weights[index[id] + offset];
+    }
+    // TODO we only need this for the datasource file since it breaks this
+    // abstraction, but uses this index
+    auto GetOffset(const DirectionalGeometryID id, const SegmentOffset offset) const
+    {
+        return index[id] + offset;
+    }
+
+    auto GetForwardGeometry(const DirectionalGeometryID id) const
+    {
+        const auto begin = nodes.cbegin() + index.at(id);
+        const auto end = nodes.cbegin() + index.at(id + 1);
+
+        return boost::make_iterator_range(begin, end);
+    }
+
+    auto GetReverseGeometry(const DirectionalGeometryID id) const
+    {
+        return boost::adaptors::reverse(GetForwardGeometry(id));
+    }
+
+    auto GetForwardDurations(const DirectionalGeometryID id) const
+    {
+        const auto begin = fwd_durations.cbegin() + index.at(id) + 1;
+        const auto end = fwd_durations.cbegin() + index.at(id + 1);
+
+        return boost::make_iterator_range(begin, end);
+    }
+
+    auto GetReverseDurations(const DirectionalGeometryID id) const
+    {
+        const auto begin = rev_durations.cbegin() + index.at(id);
+        const auto end = rev_durations.cbegin() + index.at(id + 1) - 1;
+
+        return boost::adaptors::reverse(boost::make_iterator_range(begin, end));
+    }
+
+    auto GetForwardWeights(const DirectionalGeometryID id) const
+    {
+        const auto begin = fwd_weights.cbegin() + index.at(id) + 1;
+        const auto end = fwd_weights.cbegin() + index.at(id + 1);
+
+        return boost::make_iterator_range(begin, end);
+    }
+
+    auto GetReverseWeights(const DirectionalGeometryID id) const
+    {
+        const auto begin = rev_weights.cbegin() + index.at(id);
+        const auto end = rev_weights.cbegin() + index.at(id + 1) - 1;
+
+        return boost::adaptors::reverse(boost::make_iterator_range(begin, end));
+    }
+
+    auto GetNumberOfSegments() const
+    {
+        return fwd_weights.size();
+    }
+
+    friend void io::read<UseShareMemory>(const boost::filesystem::path &path,
+                                         detail::SegmentDataContainerImpl<UseShareMemory> &segment_data);
+    friend void io::write<UseShareMemory>(const boost::filesystem::path &path,
+                                          const detail::SegmentDataContainerImpl<UseShareMemory> &segment_data);
+
+  private:
+    Vector<std::uint32_t> index;
+    Vector<NodeID> nodes;
+    Vector<EdgeWeight> fwd_weights;
+    Vector<EdgeWeight> rev_weights;
+    Vector<EdgeWeight> fwd_durations;
+    Vector<EdgeWeight> rev_durations;
+};
+}
+
+using SegmentDataView = detail::SegmentDataContainerImpl<true>;
+using SegmentDataContainer = detail::SegmentDataContainerImpl<false>;
```

Comment:
Btw what we should do (not only here) is to stick to enums and specialize based on the enums instead of true / false.

    SegmentDataContainerImpl<false> // ?? what is false? I have to look this up
    SegmentDataContainerImpl<DoNotUseSharedMemory> // ah!